### ISSUE TITLE
Chapter 5-3 giving some texture bug

### ISSUE THREAD

**Issue by RiasatSalminSami** [nipkownix/re4_tweaks#126]
![Resident Evil 4   Biohazard 4 Screenshot 2022 02 18 - 18 11 58 29](https://user-images.githubusercontent.com/50959173/154682310-7c2515c7-b2c1-47f4-baa3-8d793de97049.png)
![Resident Evil 4   Biohazard 4 Screenshot 2022 02 18 - 18 11 07 22](https://user-images.githubusercontent.com/50959173/154682318-78f18a87-d45c-46a5-8895-488b02f8e36b.png)

Everything was working fine until I killed "It" and went to next location. First my Red 9 bugged out, followed by merchant and then krauser. I'll try to recreate the bug and see if it repeats. 

Fortunately I was recording the game when it happened, so I'm gonna link the whole footage here [Video is still uploading, will take a while to finish uploading and process]:
https://youtu.be/5T5-X2nCt7k

Time stamp: 27:27

Actually just realized the bug appeared immediately after the cutscene showing death of "it", the first gun to bug out was Striker after I equipped it.

**Comment by nipkownix** [comment]
Hmm, weird issue. I imagine it doesn't happen without the HD project installed, right? Might be something wrong with that cutscene you mentioned. Will ask Albert if he knows anything about it.

**Comment by Mister-Curious** [comment]
This is a memory buffer issue that happens from time to time. Make you sure you have the 4B patch applied and that you have enough video memory

**Comment by Mister-Curious** [comment]
I have a RTX 2070 super with 8gb ram and lots of system ram and even then I get this issue from time to time, but this is usually with very taxing custom rooms I have made.

**Comment by RiasatSalminSami** [comment]
> I have a RTX 2070 super with 8gb ram and lots of system ram and even then I get this issue from time to time, but this is usually with very taxing custom rooms I have made.

I realzied the cutscene models were loading just fine after watching my own footage again.

So this is not a memory issue then? I wasn't playing on custom rooms.

**Comment by emoose** [comment]
Hmm, since it only happened after inventory was opened I wonder if it's related to RaiseInventoryAlloc.. the pG+0x3C stuff we had problems with was part of code that'd load in fonts after all.

If anyone can find a way to force this to happen it might be interesting to look at it with PIX graphics debugger (even a custom mod that it happens with could be good to check with)

**Comment by emoose** [comment]
More font texture weirdness (don't know what version re4_tweaks user had though): https://drive.google.com/file/d/1ePh31TfOIgVbGB-GCcn3ygPaWLmR7qS5/view

**Comment by nipkownix** [comment]
Huh. Weird one.

Here's what albert said about the images in the OP:

> Leon's face is now stored in the 07000000.pack file which is loaded at the beginning of the game and always in memory.
> From the look of the pictures, it seems a bug related to textures called from Em folder models (all textures stored inside 000000xx.pack, 010000xx.pack and 020000xx.pack files (enemies, characters and weapons ingame models))

It does look like the game simply stops loading these pack files at some point. If you don't know, the reason every texture turns into a font file is because that's what the game seems to consider a "default" texture. I remember having that show up when .pack and .lfs tools first became available and I decided to mess with textures for fun.

>(I think font textures are loaded with GXLoadImagePack stuff which seems seperate to all the heap crap though, so maybe it's not actually related to it...)

Raz0r's dll actually changes something related to GXLoadImagePack, but I don't remember what. 

>Also that reminds me, maybe we should display re4_tweaks version on the UI somewhere, so users are more likely to mention it.

True. I'll add it in the raw-input branch, since the logging there has the necessary stuff for it now.

**Comment by emoose** [comment]
Could be related to the 256 texture limit I saw, if you open the IDA 7.7 database and look at local types I added a GX_D3DState struct there, which has a member `IDirect3DTexture9 *CurTexture_C[256];`
Kinda looks like only 256 textures can be rendered each frame, who knows if that's actually being reached or not though.

Extending that would probably take some work, I think there's some other stuff in GX_D3DState that describe each texture too (eg. look at `GXExt_GXTex_to_D3DTex`)

Maybe will try looking into some way of displaying texture count on-screen, might be helpful for next time someone runs into it.

**Comment by Mister-Curious** [comment]
_"font file is because that's what the game seems to consider a "default" texture (it is also inside 07000000.pack, which is always in memory)."_
Actually the default texture (the black font texture is located in 0f000001.pack which is also always loaded into memory. 

_"Could be related to the 256 texture limit "_ I can say for sure that this happens without exceeding 256 textures, as I have a custom room that this issue happens in frequently (the issue of the handgun missing textures, and the muzzle fire missing textures). 

I definitely think this one is a memory issue because I had the exact same problem  (the issue of the handgun missing textures, and the muzzle fire missing textures) was fixed when I reduced the amount of water effects being drawn. I had over 20 instances of a water effect in a room (each using 60 textures), being loaded at once. The room was a bit unstable and after some troubleshooting I found that the issue was related to the water effect. What I did to reduce the load was to create an EAR triggerzone to only allow visible instances of the water effect to be drawn.

**Comment by emoose** [comment]
Yeah seems I was wrong about the 256 limit, not sure what that was for, maybe some kind of effect or something.

Found a more likely limit though, `GXInitTexObj` etc makes use of an array at 0x10C8008 to store info about each texture, each entry is 0x30 bytes, and array is 0x23280 bytes, leaving room for 0xBB8 / 3000 textures.

The uint32 at 0x10ECB84 seems to be number of total allocated textures, seems this only increases and never gets lowered in games lifetime (but I think game still reuses freed textures, at least it seems to scan for freed ones), but my guess is corruption probably happens when that goes above 3000.

Might be possible to increase the buffer for that, seems to only have a couple usages at least.

**Comment by nipkownix** [comment]
@Mister-Curious is there a chance you could provide us a test map that triggers this issue reliably?

**Comment by Mister-Curious** [comment]
> @Mister-Curious is there a chance you could provide us a test map that triggers this issue reliably?

Yes I can have that ready for you tomorrow sometime. Shoot me your email.

**Comment by emoose** [comment]
@Mister-Curious not sure if you've sent to nipkow already, but could you mail me that level too? email is on profile page https://github.com/emoose

**Comment by Mister-Curious** [comment]
@emoose sent you an email.

**Comment by nipkownix** [comment]
Oh, sorry, I missed your message, @Mister-Curious. Never had this many issues in a repo before, :p

Here's my email: skidrowa@gmail.com

**Comment by emoose** [comment]
@Mister-Curious replied, my email likely got sent to spam folder though, maybe I need to look into switching email host...
